File: src/proxy/seccomp.cpp
/* inspired by fastly/h2o:lib/common/seccomp.c @ 3edddbdb73debeedcfb78b4f4154eb9319be7d06 */
#define __unused __attribute__((unused))

#define SECCOMP_AUDIT_ARCH AUDIT_ARCH_X86_64

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/resource.h>
#include <sys/prctl.h>
#include <sys/mman.h>
#include <sys/syscall.h>

#include <linux/net.h>
#include <linux/audit.h>
#include <linux/filter.h>
#include <linux/seccomp.h>
#include <elf.h>

#include <asm/unistd.h>

#include <sysexits.h>
#include <errno.h>
#include <assert.h>
#include <err.h>
#include <signal.h>
#include <stdarg.h>
#include <stddef.h>  /* for offsetof */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>

#include "seccomp.h"

#if defined(__has_feature)
  #if __has_feature(address_sanitizer)
    #define IS_ASAN_BUILD 1
  #endif
#elif defined(__SANITIZE_ADDRESS__)
  #define IS_ASAN_BUILD 1
#endif

#if IS_ASAN_BUILD
#include <sys/ptrace.h>
#endif

static int seccomp_initialized;

#define SECCOMP_FILTER_FAIL SECCOMP_RET_TRAP

// https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/x86/entry/syscalls/syscall_64.tbl?h=v5.4.143#n343
#if !defined(__NR_statx) && defined(__linux__) && defined(__x86_64__)
#define __NR_statx 332
#endif

static const struct sock_filter main_insns[] = {
    BPF_STMT(BPF_LD+BPF_W+BPF_ABS,
        offsetof(struct seccomp_data, arch)),
    BPF_JUMP(BPF_JMP+BPF_JEQ+BPF_K, SECCOMP_AUDIT_ARCH, 1, 0),
    BPF_STMT(BPF_RET+BPF_K, SECCOMP_FILTER_FAIL),
    BPF_STMT(BPF_LD+BPF_W+BPF_ABS,
        offsetof(struct seccomp_data, nr)),

    SC_ALLOW(__NR_gettid),
    SC_ALLOW(__NR_sched_yield),
    SC_ALLOW(__NR_sched_getaffinity),
    SC_ALLOW(__NR_sched_getparam),
    SC_ALLOW(__NR_sched_getscheduler),
    SC_ALLOW(__NR_sched_setscheduler),
    SC_ALLOW(__NR_sigaltstack),
    SC_ALLOW(__NR_wait4),
    SC_ALLOW(__NR_getppid),
    SC_ALLOW(__NR_dup2),
    SC_ALLOW(__NR_fork),
#if IS_ASAN_BUILD
    SC_ALLOW_ARG(__NR_ptrace, 0, PTRACE_GETREGS /* 12 */),
    SC_ALLOW_ARG(__NR_ptrace, 0, PTRACE_ATTACH /* 16 */),
    SC_ALLOW_ARG(__NR_ptrace, 0, PTRACE_DETACH /* 17 */),
    SC_ALLOW(__NR_execve), // ASAN might use the llvm-symbolizer command
#endif

    SC_ALLOW(__NR_sched_setaffinity),
    SC_ALLOW(__NR_open),
    SC_ALLOW(__NR_openat),
    SC_ALLOW(__NR_getcwd),
    SC_ALLOW(__NR_fallocate),
    SC_ALLOW(__NR_unlink),
    SC_ALLOW(__NR_getdents),
    SC_ALLOW(__NR_setgid),
    SC_ALLOW(__NR_rename),

    SC_ALLOW(__NR_rt_sigaction),
    SC_ALLOW(__NR_rt_sigreturn),
    SC_ALLOW(__NR_lstat),
    SC_ALLOW(__NR_ioctl),
    SC_ALLOW(__NR_newfstatat),
    SC_ALLOW(__NR_sendmmsg),
    SC_ALLOW(__NR_poll),
    SC_ALLOW(__NR_stat),
    SC_ALLOW(__NR_fstat),
#ifdef __NR_statx
    SC_ALLOW(__NR_statx),
#endif
    SC_ALLOW(__NR_pread64),
    SC_ALLOW(__NR_getpeername),
    SC_ALLOW(__NR_epoll_wait),
    SC_ALLOW(__NR_epoll_ctl),
    SC_ALLOW(__NR_epoll_create),
    SC_ALLOW(__NR_epoll_create1),
    SC_ALLOW(__NR_eventfd2),
    SC_ALLOW(__NR_dup),
    SC_ALLOW(__NR_getdents64),

    /*
     * NB: we can probably revise this to only allow PF_UNIX sockets.
     */
    SC_ALLOW(__NR_socket),
    SC_ALLOW(__NR_lseek),
    SC_ALLOW(__NR_socketpair),
    SC_ALLOW(__NR_writev),
    SC_ALLOW(__NR_readv),
    SC_ALLOW(__NR_tgkill),
    SC_ALLOW(__NR_set_robust_list),
    SC_ALLOW(__NR_clone),
    SC_ALLOW(__NR_accept4),
    SC_ALLOW(__NR_pipe),
    SC_ALLOW(__NR_getsockopt),
    SC_ALLOW(__NR_setsockopt),
    SC_ALLOW(__NR_sendmsg),
    SC_ALLOW(__NR_recvmsg),
    SC_ALLOW(__NR_recvmmsg),
    SC_ALLOW(__NR_sendto),
    SC_ALLOW(__NR_recvfrom),
    SC_ALLOW(__NR_sysinfo),
    SC_ALLOW(__NR_fcntl),
    SC_ALLOW(__NR_prctl),
    /*
     * Dangerous NB:
     */
    SC_ALLOW(__NR_bind),
    SC_ALLOW(__NR_connect),
    SC_ALLOW(__NR_accept),
    SC_ALLOW(__NR_listen),
    SC_ALLOW(__NR_readlink),
    /*
     * Additional syscalls that were smoked out of the CI process.
     */
    SC_ALLOW(__NR_uname),
    SC_ALLOW(__NR_access),
    SC_ALLOW(__NR_setuid),
    SC_ALLOW(__NR_setgroups),
    SC_ALLOW(__NR_getsockname),
    SC_ALLOW(__NR_setpriority),
    SC_ALLOW(__NR_getpriority),

    SC_ALLOW(__NR_brk),
    SC_ALLOW(__NR_clock_gettime),
    SC_ALLOW(__NR_close),
    SC_ALLOW(__NR_exit),
    SC_ALLOW(__NR_exit_group),
    SC_ALLOW(__NR_futex),
    SC_ALLOW(__NR_geteuid),
    SC_ALLOW(__NR_getpgid),
    SC_ALLOW(__NR_getpid),
    SC_ALLOW(__NR_getrandom),
    SC_ALLOW(__NR_gettimeofday),
    SC_ALLOW(__NR_getuid),
    SC_ALLOW(__NR_madvise),
    SC_ALLOW(__NR_restart_syscall),
    SC_ALLOW(__NR_getrusage),
    /*
     * Permit mmap(2) and mprotect(2) operations, but do not allow h2o
     * to create/update mappings with PROT_EXEC. This will make it harder
     * to overwrite the various callbacks in h2o to point at something
     * exploit defined.
     */
    //SC_ALLOW_ARG_MASK(__NR_mmap, 3, PROT_READ|PROT_WRITE|PROT_NONE),
    SC_ALLOW(__NR_mmap),
    SC_ALLOW_ARG_MASK(__NR_mprotect, 2, PROT_READ|PROT_WRITE|PROT_NONE),
    SC_ALLOW(__NR_mremap),
    SC_ALLOW(__NR_munmap),
    SC_ALLOW(__NR_nanosleep),
    SC_ALLOW(__NR_clock_nanosleep),
    SC_ALLOW(__NR_read),
    SC_ALLOW(__NR_rt_sigprocmask),
    SC_ALLOW(__NR_select),
    SC_ALLOW(__NR_shutdown),
    SC_ALLOW(__NR_time),
    SC_ALLOW(__NR_write),
    SC_ALLOW(__NR_bpf),
    SC_ALLOW(__NR_inotify_init1),
    SC_ALLOW(__NR_inotify_add_watch),
    SC_ALLOW(__NR_inotify_rm_watch),
    SC_ALLOW(__NR_clock_getres),
    BPF_STMT(BPF_RET+BPF_K, SECCOMP_FILTER_FAIL),
};

static const struct sock_fprog basic_program = {
    .len = (sizeof(main_insns) / sizeof(main_insns[0])),
    .filter = (struct sock_filter *)main_insns,
};

static void sandbox_violation(__unused int signum, siginfo_t *info, __unused void *void_context)
{

    (void) fprintf(stdout,
        "%s: unexpected system call (arch:0x%x,syscall:%d @ %p)\n",
        __func__, info->si_arch, info->si_syscall, info->si_call_addr);
    abort();
}

static void seccomp_helper(void)
{
    struct sigaction act;
    sigset_t mask;

    memset(&act, 0, sizeof(act));
    sigemptyset(&mask);
    sigaddset(&mask, SIGSYS);
    act.sa_sigaction = &sandbox_violation;
    act.sa_flags = SA_SIGINFO;
    if (sigaction(SIGSYS, &act, NULL) == -1) {
        err(1, "sigaction(SIGSYS) failed");
    }
    if (sigprocmask(SIG_UNBLOCK, &mask, NULL) == -1) {
        err(1, "sigprocmask(SIGSYS) failed");
    }
}

void seccomp_bind(void)
{
    const struct sock_fprog *fprog;

    if (!seccomp_initialized) {
        seccomp_helper();
        seccomp_initialized = 1;
    }
    fprog = &basic_program;
    if (prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0) == -1) {
        err(1, "prctl(PR_SET_NO_NEW_PRIVS)");
    }
    if (prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, fprog) == -1) {
        err(1, "prctl(PR_SET_SECCOMP)");
    }
    printf("[SECCOMP] security policy bound to thread\n");
}

File: src/core/seccomp.rs
//! This module provides Rust definitions of macros and structs used for BPF programs.
//! Additionally, some definitions are included that are seccomp-specific.
//!
//! ### Why not use the `seccomp` crate?
//!
//! The `seccomp` crate is built on `libseccomp`, is fairly unmaintained, and doesn't use bindgen
//! to bind underlying definitions either. Meanwhile, we only need a few definitions for ExecuteD's
//! seccomp program, so for now this is a sufficient stand-in.

// Some definitions, like BPF filter code constants, here might not get used today, but are
// included for completeness.
#![allow(dead_code)]

pub mod syscalls;

use log::{error, trace};
use nix::errno::Errno;
use std::convert::TryInto;
use std::sync::Mutex;

lazy_static::lazy_static! {
    static ref SECCOMP_ACTIVE: Mutex<bool> = Mutex::new(false);
}

macro_rules! SECCOMP_ALLOW {
    ( $($x:expr),* ) => {
        {
        let mut temp_vec = Vec::new();
        $(
        temp_vec.push(FilterInstruction::load_nr());
        temp_vec.push(FilterInstruction::jump(FilterCode::cmp_eq_const(), $x, 0, 1));
        temp_vec.push(FilterInstruction::allow());
        )*
        temp_vec
        }
    };
}

pub fn install_seccomp_connect_filter() {
    use self::syscalls::*;
    let mut syscall_allow_list = SECCOMP_ALLOW!(
        __NR_ACCEPT4,
        __NR_BIND,
        __NR_BRK,
        __NR_CLOCK_GETTIME,
        __NR_CLOCK_NANOSLEEP,
        __NR_CLONE,
        __NR_CLONE3,
        __NR_CLOSE,
        __NR_CONNECT,
        __NR_EPOLL_CREATE,
        __NR_EPOLL_CREATE1,
        __NR_EPOLL_CTL,
        __NR_EPOLL_WAIT,
        __NR_EVENTFD2,
        __NR_EXIT,
        __NR_EXIT_GROUP,
        __NR_FADVISE64,
        __NR_FCNTL,
        __NR_FSTAT,
        __NR_NEWFSTATAT,
        __NR_FUTEX,
        __NR_GET_ROBUST_LIST,
        __NR_GETCWD,
        __NR_GETDENTS,
        __NR_GETDENTS64,
        __NR_GETPEERNAME,
        __NR_GETPID,
        __NR_GETPPID,
        __NR_GETRANDOM,
        __NR_GETSOCKNAME,
        __NR_GETSOCKOPT,
        __NR_GETTID,
        __NR_GETTIMEOFDAY,
        __NR_GETUID,
        __NR_INOTIFY_INIT,
        __NR_INOTIFY_INIT1,
        __NR_INOTIFY_ADD_WATCH,
        __NR_INOTIFY_RM_WATCH,
        __NR_IOCTL,
        __NR_KILL,
        __NR_LISTEN,
        __NR_LSEEK,
        __NR_LSTAT,
        __NR_MADVISE,
        __NR_MMAP,
        __NR_MPROTECT,
        __NR_MREMAP,
        __NR_MUNMAP,
        __NR_NANOSLEEP,
        __NR_OPEN,
        __NR_OPENAT,
        __NR_PIPE2,
        __NR_POLL,
        __NR_PPOLL,
        __NR_PRCTL,
        __NR_PREAD64,
        __NR_READ,
        __NR_READLINK,
        __NR_RECVFROM,
        __NR_RECVMSG,
        __NR_RESTART_SYSCALL,
        __NR_RSEQ,
        __NR_RT_SIGACTION,
        __NR_RT_SIGPROCMASK,
        __NR_RT_SIGRETURN,
        __NR_SCHED_GETAFFINITY,
        __NR_SCHED_GETPARAM,
        __NR_SCHED_GET_PRIORITY_MAX,
        __NR_SCHED_GETSCHEDULER,
        __NR_SCHED_SETSCHEDULER,
        __NR_SCHED_YIELD,
        __NR_SENDFILE,
        __NR_SENDTO,
        __NR_SET_ROBUST_LIST,
        __NR_SETSOCKOPT,
        __NR_SHUTDOWN,
        __NR_SIGALTSTACK,
        __NR_SOCKETPAIR,
        __NR_STAT,
        __NR_STATFS,
        __NR_STATX,
        __NR_TGKILL,
        __NR_UNAME,
        __NR_UNLINK,
        __NR_WRITE,
        __NR_WRITEV
    );
    let socket_filter = vec![
        FilterInstruction::load_nr(),
        FilterInstruction::jump(FilterCode::cmp_eq_const(), __NR_SOCKET, 1, 0),
        /* socket is the last syscall we check, if this isn't it, kill the process */
        FilterInstruction::trap(),
        FilterInstruction::load_arg(0), // check that the domain is `UNIX`
        FilterInstruction::jump(
            FilterCode::cmp_eq_const(),
            nix::sys::socket::AddressFamily::Unix as u32,
            0, // if this is a unix socket request, fall through to allow
            1, // else branch to errno
        ),
        FilterInstruction::allow(),
        FilterInstruction::errno(libc::EPERM as u16),
    ];
    syscall_allow_list.extend(socket_filter);
    let load_res = load(&syscall_allow_list);
    if let Err(e) = load_res {
        error!("failed to install seccomp filter: {}", e);
        std::process::exit(1);
    }
    trace!("loaded socket(2) filter, non-Unix sockets are now denied");

    #[cfg(debug_assertions)]
    {
        match std::net::TcpStream::connect("127.0.0.1:80").err() {
            Some(err) => {
                if err.kind() != std::io::ErrorKind::PermissionDenied {
                    panic!("unexpected error exercising seccomp filter: {}", err);
                }
                // else we got a PermissionDenied, like we wanted
            }
            None => {
                panic!("erroneously able to create a TCP socket after installing seccomp filter to deny that permission");
            }
        }
    }
}

fn activate_seccomp() -> Result<(), Errno> {
    let activation_result = unsafe { libc::prctl(libc::PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0) };
    if activation_result != 0 {
        Err(Errno::last())
    } else {
        Ok(())
    }
}

/// If the program failed to load, return a `false`. There will be a value in
fn load(program: &[FilterInstruction]) -> Result<(), Errno> {
    let mut seccomp_active_guard = SECCOMP_ACTIVE.lock().unwrap();
    if !*seccomp_active_guard {
        let seccomp_activated = activate_seccomp();
        if seccomp_activated.is_err() {
            trace!("failed to activate seccomp");
            return seccomp_activated;
        }
        trace!("seccomp activated");
        *seccomp_active_guard = true;
    }
    drop(seccomp_active_guard);

    let program = FilterProgram {
        len: program
            .len()
            .try_into()
            .expect("program is smaller than 65535 instructions"),
        filter: program.as_ptr(),
    };
    let load_result = unsafe {
        libc::prctl(
            libc::PR_SET_SECCOMP,
            libc::SECCOMP_MODE_FILTER,
            &program as *const _,
        )
    };
    if load_result != 0 {
        Err(Errno::last())
    } else {
        Ok(())
    }
}

/// This is Rust for `struct sock_fprog`. It's kind of like a field-swapped slice, but the length
/// is a `u16`, not `usize`.
#[repr(C)]
struct FilterProgram {
    len: u16,
    filter: *const FilterInstruction,
}

/// This is Rust for `struct sock_filter`.
#[repr(C)]
struct FilterInstruction {
    code: u16,
    jt: u8, // "Jump true"
    jf: u8, // "Jump false"
    k: u32, // "Generic multiuse field". Typically a constant. K for konstant?
}

// All taken from <linux/bpf_common.h>
const BPF_LD: u16 = 0x00;
const BPF_LDX: u16 = 0x01;
const BPF_ST: u16 = 0x02;
const BPF_STX: u16 = 0x03;
const BPF_ALU: u16 = 0x04;
const BPF_JMP: u16 = 0x05;
const BPF_RET: u16 = 0x06;
const BPF_MISC: u16 = 0x07;

// Sizes, from <linux/bpf_common.h>
const BPF_W: u16 = 0x00;
const BPF_H: u16 = 0x08;
const BPF_B: u16 = 0x10;

/// An enum form of BPF-permitted memory operation sizes. This is just to prevent invalid sizes
/// from being provided when building filter instructions.
#[derive(Copy, Clone, Debug)]
#[repr(u16)]
pub enum MemOpSize {
    W = BPF_W,
    H = BPF_H,
    B = BPF_B,
}

// Access modes, from <linux/bpf_common.h>
const BPF_IMM: u16 = 0x10;
const BPF_ABS: u16 = 0x20;
const BPF_IND: u16 = 0x40;
const BPF_MEM: u16 = 0x60;
const BPF_LEN: u16 = 0x80;
const BPF_MSH: u16 = 0xa0;

// Jump conditions, from <linux/bpf_common.h>
const BPF_JA: u16 = 0x00;
const BPF_JEQ: u16 = 0x10;
const BPF_JGT: u16 = 0x20;
const BPF_JGE: u16 = 0x30;
const BPF_JSET: u16 = 0x40;

// Source, from <linux/bpf_common.h>
const BPF_K: u16 = 0x00;
const BPF_X: u16 = 0x08;

struct FilterCode;

impl FilterCode {
    pub fn cmp_eq_const() -> u16 {
        BPF_JMP + BPF_JEQ + BPF_K
    }
    pub fn ld_abs(size: MemOpSize) -> u16 {
        BPF_LD + (size as u16) + BPF_ABS
    }
    pub fn ret() -> u16 {
        BPF_RET + BPF_K
    }
}

const SECCOMP_RET_ERRNO: u32 = 0x00050000;
const SECCOMP_RET_TRACE: u32 = 0x7ff00000;
const SECCOMP_RET_ALLOW: u32 = 0x7fff0000;
const SECCOMP_RET_TRAP: u32 = 0x00030000;
const SECCOMP_RET_KILL_THREAD: u32 = 0x00000000;
const SECCOMP_RET_KILL_PROCESS: u32 = 0x80000000;

// helper functions to construct FilterInstruction. Note that these helpers may involve
// seccomp-specific details.
impl FilterInstruction {
    /// Rust for `BPF_STMT`. Not a macro, nor a const fn, unfortunately.
    pub fn statement(code: u16, k: u32) -> Self {
        FilterInstruction {
            code,
            jt: 0,
            jf: 0,
            k,
        }
    }
    /// Rust for `BPF_JUMP`. Not a macro, nor a const fn, unfortunately. `BPF_JUMP` swaps argument
    /// order from the struct layout too.
    pub fn jump(code: u16, k: u32, jt: u8, jf: u8) -> Self {
        FilterInstruction { code, jt, jf, k }
    }
    pub fn load_arg(n: u8) -> Self {
        assert!(n < 6);

        let args_offset: u32 = memoffset::offset_of!(SeccompData, args)
            .try_into()
            .expect("args array is a reasonably small offset");
        let selected_arg = args_offset
            .checked_add((8 * n) as u32)
            .expect("arg[n] is also a reasonably small offset");
        FilterInstruction::statement(FilterCode::ld_abs(MemOpSize::W), selected_arg)
    }
    pub fn load_nr() -> Self {
        let nr_offset: u32 = memoffset::offset_of!(SeccompData, nr)
            .try_into()
            .expect("syscall nr field is a reasonably small offset");
        FilterInstruction::statement(FilterCode::ld_abs(MemOpSize::W), nr_offset)
    }
    pub fn allow() -> Self {
        FilterInstruction::statement(FilterCode::ret(), SECCOMP_RET_ALLOW)
    }
    pub fn trap() -> Self {
        FilterInstruction::statement(FilterCode::ret(), SECCOMP_RET_TRAP)
    }
    pub fn kill_process() -> Self {
        FilterInstruction::statement(FilterCode::ret(), SECCOMP_RET_KILL_PROCESS)
    }
    pub fn errno(errno: u16) -> Self {
        FilterInstruction::statement(FilterCode::ret(), SECCOMP_RET_ERRNO | (errno as u32))
    }
}

/// This is Rust for `struct seccomp_data`, the struct of information a filter operates over.
/// Translated from <linux/seccomp.h>.
#[repr(C)]
struct SeccompData {
    nr: i32,                  // The system call number
    arch: u32, // system call convention as an AUDIT_ARCH_* value defined in <linux/audit.h>
    instruction_pointer: u64, // at the time of the system call
    args: [u64; 6], // up to 6 system call arguments always stored as 64-it values regardless of the architecture.
}
